server/sonar-server/src/test/java/org/sonar/server/component/ws/ShowActionTest.java

```diff
@@ -0,0 +1,170 @@
+
+@Category(DbTests.class)
+public class ShowActionTest {
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+  @Rule
+  public UserSessionRule userSession = UserSessionRule.standalone()
+    .setGlobalPermissions(GlobalPermissions.SYSTEM_ADMIN);
+  @Rule
+  public DbTester db = DbTester.create(System2.INSTANCE);
+  ComponentDbTester componentDb = new ComponentDbTester(db);
+
+  WsActionTester ws = new WsActionTester(new ShowAction(userSession, db.getDbClient(), new ComponentFinder(db.getDbClient())));
+
+  @Test
+  public void json_example() throws IOException {
+    insertJsonExampleComponentsAndSnapshots();
+
+    String response = ws.newRequest()
+      .setParam("id", "AVIF-FffA3Ax6PH2efPD")
+      .execute()
+      .getInput();
+
+    assertJson(response).isSimilarTo(getClass().getResource("show-example.json"));
+  }
+
+  @Test
+  public void show_by_key_with_project_permission() {
+    userSession.anonymous().login().addProjectUuidPermissions(UserRole.ADMIN, "project-uuid");
+    componentDb.insertProjectAndSnapshot(newProjectDto("project-uuid").setKey("project-key"));
+
+    ShowWsResponse response = newRequest(null, "project-key");
+
+    assertThat(response.getAncestorsCount()).isEqualTo(0);
+    assertThat(response.getComponent().getKey()).isEqualTo("project-key");
+  }
+
+  @Test
+  public void show_with_browse_permission() {
+    userSession.anonymous().addProjectUuidPermissions(UserRole.USER, "project-uuid");
+    componentDb.insertProjectAndSnapshot(newProjectDto("project-uuid"));
+
+    ShowWsResponse response = newRequest("project-uuid", null);
+
+    assertThat(response.getComponent().getId()).isEqualTo("project-uuid");
+  }
+
+  @Test
+  public void show_provided_project() {
+    componentDb.insertComponent(newProjectDto("project-uuid"));
+
+    ShowWsResponse response = newRequest("project-uuid", null);
+
+    assertThat(response.getComponent().getId()).isEqualTo("project-uuid");
+  }
+
+  @Test
+  public void fail_if_not_enough_privilege() {
+    userSession.anonymous().setGlobalPermissions(GlobalPermissions.QUALITY_PROFILE_ADMIN);
+    expectedException.expect(ForbiddenException.class);
+    componentDb.insertProjectAndSnapshot(newProjectDto("project-uuid"));
+
+    newRequest("project-uuid", null);
+  }
+
+  @Test
+  public void fail_if_component_does_not_exist() {
+    expectedException.expect(NotFoundException.class);
+    expectedException.expectMessage("Component id 'unknown-uuid' not found");
+
+    newRequest("unknown-uuid", null);
+  }
+
+  private ShowWsResponse newRequest(@Nullable String uuid, @Nullable String key) {
+    TestRequest request = ws.newRequest()
+      .setMediaType(MediaTypes.PROTOBUF);
+
+    if (uuid != null) {
+      request.setParam(PARAM_ID, uuid);
+    }
+    if (key != null) {
+      request.setParam(PARAM_KEY, key);
+    }
+
+    InputStream responseStream = request.execute().getInputStream();
```

Comment:
could be `try(InputStream responseStream = request.execute().getInputStream()) {`